### ISSUE TITLE
jshint error for 'describe/it not defined' on generated helper tests

### ISSUE THREAD

**Issue by DisruptiveMind** [switchfly/ember-cli-mocha#37]
When generating a new helper the associated test does not run due to globals `'describe'` and `'it'` not being defined.

**Generated**
```javascript
/* jshint expr:true */
import {
  aHelper
} from 'hub/helpers/a-helper';

describe('AHelperHelper', function() {
  // Replace this with your real tests.
  it('works', function() {
    var result = aHelper(42);
    expect(result).to.be.ok;
  });
});
```

Results in: 
```javascript
unit/helpers/a-helper-test.js: line 6, col 1, 'describe' is not defined.
unit/helpers/a-helper-test.js: line 8, col 3, 'it' is not defined.
```

**solution....sort of**
I manually added `'describe'` and `'it'` to ***tests/.jshintrc*** after looking at #3 

Is there another way of doing this or am I hooking things up wrong somewhere else?

**Comment by jonathanKingston** [comment]
:+1: Would appreciate the correct path to import 'describe' from. The generators here don't specify one.

**Comment by DisruptiveMind** [comment]
I just added **describe** & **it** inside ***tests/.jshintrc*** and it seems to have solved it for now.
I can `generate helper helper-name` without the generated tests tripping JSHint now.

**tests/.jshintrc**
```javascript
{
  ⋮
  "predef": [
    ⋮
    "describe",
    "it"
    ⋮
  ],
  ⋮
}
```

**Comment by jonathanKingston** [comment]
Ember-cli 0.1.13 update is trying to move away from any global like this: http://reefpoints.dockyard.com/2015/02/06/ember-qunit-0-2.html

https://github.com/ember-cli/ember-cli/releases/tag/v0.1.13

**Comment by rwjblue** [comment]
I think this should be handled up stream, and created an issue in ember-mocha for it: https://github.com/switchfly/ember-mocha/issues/22

**Comment by jonathanKingston** [comment]
@rwjblue good call, thanks for raising.

Is it worth raising about exposing `assert`, `should` and `expects` from chai.js too?

**Comment by rwjblue** [comment]
> Is it worth raising about exposing assert, should and expects from chai.js too?

Not sure, I'm not a heavy ember-mocha user, I was mostly trying to help point @dgeb and others towards a possible solution (that we pulled into ember-qunit).

**Comment by dgeb** [comment]
Thanks @rwjblue 

@jonathanKingston @DisruptiveMind et al - PRs welcome over in ember-mocha as spec'd out in https://github.com/switchfly/ember-mocha/issues/22

> Is it worth raising about exposing assert, should and expects from chai.js too?

Seems like an additional top-level `chai` module would be most appropriate.

**Comment by jonathanKingston** [comment]
@dgeb I'll make a PR for the describe at some point soonish unless someone else gets to it first.

I made a quick chai package to expose the main interfaces to chai. This might be a better fragmentation from just having ember-cli module define the use of Chai too: https://github.com/jonathanKingston/ember-cli-chai

**Comment by rwjblue** [comment]
Submitted https://github.com/switchfly/ember-mocha/pull/23 to add the shims, this repo will still need to be updated (mostly blueprints I think) once that lands.

**Comment by DisruptiveMind** [comment]
I've just added the import statements for the helpers blueprint in disruptivemind/ember-cli-mocha@afe57116b35d5659efde57aff8b3d98ac7f1e841. I'm holding off the PR for now.

Just wondering, should `expect` be imported as well?

**Comment by jonathanKingston** [comment]
@DisruptiveMind I was thinking about this the other day, would it make more sense to have ember-cli-mocha more configurable so that any assertion library can be used. The addon could take a config argument for example, which would then modify the helpers accordingly.

So something like: `config/environment.js`:
```
ENV['ember-cli-mocha'] = {
  assertion: 'chai.assert' // Valid values: 'chai.assert', 'chai.should', 'chai.expects', 'qunit'
}
```

Would generate:
```
/* jshint expr:true */
 import {
  describe,
  it
} from 'ember-mocha';

import {
  assert
} from 'chai';
```

This might be something that can be made generic enough to be usable from ember-cli-qunit too.

What do you think @rwjblue.

**Comment by DisruptiveMind** [comment]
@jonathanKingston :+1: on configurable options

**Comment by dgeb** [comment]
Please upgrade to ember-cli-mocha 0.5.0 and then run `ember g ember-cli-mocha` to get the latest ember-mocha.

ember-mocha now exports shimmed [mocha](https://github.com/switchfly/ember-mocha/blob/master/lib/mocha.js) and [chai](https://github.com/switchfly/ember-mocha/blob/master/lib/chai.js) modules.

Note that only `expect` and `assert` are exported from `chai` for now. `should` requires injections which most developers will want to avoid (perhaps we can figure out how to better support `should` at some point).

I've also updated the blueprints to use these new shims. I like the idea in theory of making these blueprints more flexible as you've described above, as long as we can do so without too much noise in the blueprints. If you want to pursue an approach, please open a new issue / PR.

**Comment by jonathanKingston** [comment]
@dgeb :+1: 

Does doing the following with 'should' not work as expected:

```
  import should from 'ember-mocha';
  should();

  thing.should...
```

As far as I was aware exporting the should method then using this within a module scope should have a self contained impact.

**Comment by rwjblue** [comment]
@jonathanKingston - I am not an expert on chai but it seems that `should` actually adds a bunch of methods to the core object prototype which means using it anywhere would affect all the rest of the tests.

**Comment by rwjblue** [comment]
https://github.com/chaijs/chai/blob/master/chai.js#L3782-L3786

**Comment by rwjblue** [comment]
But I do agree that the `chai` shim should export `should`, and the actual assertion lib/mechanism is completely independent (as it should be).

**Comment by jonathanKingston** [comment]
@rwjblue yeah it does, I guess I just assumed ES6 modules completely isolated their code even for native prototypes (It hasn't been something I had tried before because of it's edge usecase - however would certainly help if that was the expected behaviour and Babel just needed to implement).

**Comment by rwjblue** [comment]
@jonathanKingston - I'm pretty sure that ES6 does not suggest that modifications to core prototypes are isolated within each module (and if it did, I highly doubt that it could be polyfilled via babel).

**Comment by dgeb** [comment]
Yeah, once `should` is invoked, its core object injections could affect other tests - even those that don't use `should`. I guess I see `should` as a footgun (i.e. anti-pattern), especially when presented alongside `expect` and `assert` as equivalents. I really am on the fence about exposing it from the `chai` shim - but I'd be willing to do so if there's sufficient demand.

**Comment by rwjblue** [comment]
> I guess I see should as a footgun (i.e. anti-pattern), especially when presented alongside expect and assert as equivalents.

This is exactly why I left it out of my initial `chai` shim PR.  I think if you *really really* want to use should syntax you can figure it out on you own.  I would personally rather steer folks towards the right way...

**Comment by rwjblue** [comment]
And, I still haven't heard anyone say they wanted to use `should` syntax: unless I am mis-reading @jonathanKingston's comment he was mostly advocating for it to be pluggable (which it is already).

**Comment by dgeb** [comment]
> And, I still haven't heard anyone say they wanted to use should syntax

@rwjblue oh, I almost forgot about #14. Let me comment there.

**Comment by jonathanKingston** [comment]
@rwjblue yes I am advocating for it to be pluggable (I was just trying to ascertain if it would be isolated).

---

From my quick scrambling through the ES6 spec (I'm probably not reading this right).

[Object Environment Records](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-object-environment-records)
> Immutable bindings do not exist for object environment records.

So no help there, however I believe that is just describing normal behaviour of scripts.


---


[Well known intrinsic objects](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-well-known-intrinsic-objects)

>  Unless otherwise specified each intrinsic object actually corresponds to a set of similar objects, one per Realm.

[Code Realms](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-code-realms)

> Before it is evaluated, all ECMAScript code must be associated with a Realm. Conceptually, a realm consists of a set of intrinsic objects, an ECMAScript global environment

[Module evaluation](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-moduleevaluation)
> Set the Realm of moduleCxt to module.[[Realm]].

---

My reading of that is `Object` is considered an *Intrinsic* object which is related to the *Realm* it is in, Modules appear to run in their own *Realm*. Creating a new *Real* appears to reinstantiate new Intrinsics.

I could be wrong in how that is supposed to be read, however if that is the case Babel could be updated with this behaviour. 

@dgeb I was just about to quote #14 :+1:

**Comment by jonathanKingston** [comment]
Also:
[Crossing realms with symbols](http://www.2ality.com/2014/12/es6-symbols.html#crossing_realms_with_symbols)

>  The problem is that each realm has its own local copy of Array and, because objects have individual identities, those local copies are considered different

So I will double check this issue exists in Babel, if it does then hopefully that can be fixed there to isolate the prototypes.
